crates/sui-core/src/authority/execution_time_estimator.rs

```diff
@@ -0,0 +1,779 @@
+
+// If our current local observation differs from the last one we shared by more than
+// this percent, we share a new one.
+const OBSERVATION_SHARING_DIFF_THRESHOLD: f64 = 0.05;
+
+// Collects local execution time estimates to share via consensus.
+pub struct ExecutionTimeObserver {
+    epoch_store: Weak<AuthorityPerEpochStore>,
+    consensus_adapter: Box<dyn SubmitToConsensus>,
+
+    local_observations: HashMap<ExecutionTimeObservationKey, LocalObservations>,
+}
+
+#[derive(Debug, Clone)]
+pub struct LocalObservations {
+    moving_average: SingleSumSMA<Duration, u32, LOCAL_OBSERVATION_WINDOW_SIZE>,
+    last_shared: Option<Duration>,
+}
+
+// Tracks local execution time observations and shares them via consensus.
+impl ExecutionTimeObserver {
+    pub fn spawn(
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+        consensus_adapter: Box<dyn SubmitToConsensus>,
+        channel_size: usize,
+    ) {
+        if epoch_store
+            .protocol_config()
+            .per_object_congestion_control_mode()
+            != PerObjectCongestionControlMode::ExecutionTimeEstimate
+        {
+            info!("ExecutionTimeObserver disabled because per-object congestion control mode is not ExecutionTimeEstimate");
+            return;
+        }
+
+        let (tx_local_execution_time, mut rx_local_execution_time) = mpsc::channel(channel_size);
+        epoch_store.set_local_execution_time_channel(tx_local_execution_time);
+
+        // TODO: pre-populate local observations with stored data from prior epoch.
+        let mut observer = Self {
+            epoch_store: Arc::downgrade(epoch_store),
+            consensus_adapter,
+            local_observations: HashMap::new(),
+        };
+        tokio::spawn(async move {
+            while let Some((tx, timings, total_duration)) = rx_local_execution_time.recv().await {
+                observer
+                    .record_local_observations(&tx, &timings, total_duration)
+                    .await;
+            }
+            info!("shutting down ExecutionTimeObserver");
+        });
+    }
+
+    #[cfg(test)]
+    fn new_for_testing(
+        epoch_store: Arc<AuthorityPerEpochStore>,
+        consensus_adapter: Box<dyn SubmitToConsensus>,
+    ) -> Self {
+        Self {
+            epoch_store: Arc::downgrade(&epoch_store),
+            consensus_adapter,
+            local_observations: HashMap::new(),
+        }
+    }
+
+    // Used by execution to report observed per-entry-point execution times to the estimator.
+    // Updates moving averages and submits observation to consensus if local observation differs
+    // from consensus median.
+    // TODO: Consider more detailed heuristic to account for overhead outside of commands.
+    async fn record_local_observations(
+        &mut self,
+        tx: &ProgrammableTransaction,
+        timings: &[ExecutionTiming],
+        total_duration: Duration,
+    ) {
+        assert_eq!(tx.commands.len(), timings.len());
```

Comment:
I missed this before - this assert is not valid. Timings can be shorter due to an abort

Reply:
Good catch, thanks